content/blog/2019-10-28-self/index.md

```diff
@@ -0,0 +1,176 @@
+  programming languages & computer architecture. In his free time, he
+  [subtweets](https://twitter.com/notypes/status/1170037148290080771) his advisor and [annoys tenured professors](https://twitter.com/natefoster/status/1074401015565291520).
+"""
+[[extra.authors]]
+name = "Rachit Nigam"
+link = "https://rachitnigam.com"
++++
+
+[An Efficient Implementation of Self, a Dynamically-Typed Object-Oriented
+Language Based on Prototypes][paper] presents techniques for runtime
+compilation, now more commonly referred to as Just In Time (JIT) compilation,
+for the [Self][] programming language.
+
+> We have developed and implemented techniques that double the performance of
+> the dynamically-typed object oriented languages.
+
+
+## The Challenge
+
+The term "dynamic language" is mostly commonly associated with modern scripting
+languages like Python and JavaScript. Self, a much older language developed
+at the famed Xerox PARC labs takes the philosophy of dynamism to a logical
+extreme -- *everything* in self is a message to an object. This includes
+Java-like method calls on object *as well as* control structures like loops
+and conditionals.
+
+For example, the Self condition `IfTrue:IfFalse:` is a method invocation on
+the boolean object `true` and `false`. This means that a simple conditional:
+
+```
+if (x) 1 else 2
+```
+
+in Self parlance (written with JavaScript-esque syntax) looks like this:
+
+
+```
+x.If({ true: () => 1, false: () => 2 })
+```
+
+The runtime behavior of this program is invoking the method `IfTrue` on the
+object `x`, which can be a `true` or `false` but is not required to be those,
+and execute the "thunk" (a function with no argument) corresponding to the true
+or the false branch.  Note that `x` is not required to be `true` or `false`.
+Any Self object can define the `If` method and specify its conditional
+behavior.
+
+Considering this dynamism, a Self compiler must:
+
+1. **Respect the semantics of method calls**. Restricting or specializing conditionals and primitives don't follow the spirit of the language.
+2. **Provide interactive speed**. Recompiling after a small change is not an option because Self is meant for rapid exploration in a programming environment.
+3. **Preserve stack traces**. Self supports extreme reflection and introspection. A programming environment must be debuggable.
+4. **Generate fast code**. Interpreting the whole language is probably too slow.
+
+## The Solution
+
+While the paper goes into the nitty-gritty of object layouts and method
+invocations, the essence of the paper can be summarized as:
+
+>  When type information for an object is available, generate specialized code
+>  and let inlining and compiler optimizations work their magic.
+
+The ideas described in the paper have influenced the design of modern JITs for
+JavaScript which use a lot of [similar features][jit] to improve JavaScript execution
+times.
```

Comment:
Not just JavaScript compilers—basically every JIT in use today descends from this set of ideas. (Even ones for more static-seeming languages like Java and C#.)